File: back/src/domain/use_cases/fetch_titles.rs
use crate::domain::models::Title;
use crate::infra::{title_repository::FetchAllError, title_repository::TitleRepository};
use std::sync::Arc;

pub enum Error {
    Unknown(&'static str),
}

pub async fn fetch_titles(repo: Arc<dyn TitleRepository>) -> Result<Vec<Title>, Error> {
    match repo.fetch_all().await {
        Ok(titles) => Ok(titles),
        Err(FetchAllError::Unknown(str)) => Err(Error::Unknown(str)),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::infra::InMemoryTitleRepository;

    #[tokio::test]
    async fn it_should_return_an_unknown_error_when_an_unexpected_error_happens() {
        let repo = Arc::new(InMemoryTitleRepository::new().with_error());

        let res = fetch_titles(repo).await;

        match res {
            Err(Error::Unknown(_)) => {}
            _ => unreachable!(),
        };
    }

    #[tokio::test]
    async fn it_should_return_all_locations_otherwise() {
        let title = Title::test();
        let repo = Arc::new(InMemoryTitleRepository::new());
        repo.insert(title.clone());

        let res = fetch_titles(repo).await;

        match res {
            Ok(titles) => {
                assert_eq!(vec![title], titles);
            }
            _ => unreachable!(),
        };
    }
}

File: src/app/api/fetchTitles.ts
import { Context } from "hono";
import * as lib from "@domain";
import { TitleRepository } from "@infra";
import { Result } from "@utils";

export async function fetchTitles(
    c: Context,
    titleRepo: TitleRepository
) {
    return Result.match(await lib.fetchTitles(titleRepo), {
        onOk: (titles) => c.json(titles.map(title => ({
            title: lib.Title.toString(title)
        }))),
        onErr: () => c.json([], 500)
    });
}
